numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -302,9 +302,42 @@ static int
             oop.real = NPY_NAN;
             oop.imag = NPY_NAN;
         }
+        else if (PyBytes_Check(op) || PyUnicode_Check(op)) {
+            /*
+             * Unlike most numeric conversion functions PyComplex_AsCComplex
+             * does not handle strings, so we have to use its constructor.
+             */
+            PyObject *pycomplex, *args;
+            if (PyBytes_Check(op)) {
+                /* The complex constructor expects unicode */
+                PyObject *unicode;
+                unicode = PyUnicode_FromEncodedObject(op, NULL, NULL);
+                if (unicode == NULL) {
+                    return -1;
+                }
+                args = PyTuple_Pack(1, unicode);
+                Py_DECREF(unicode);
+            }
+            else {
+                args = PyTuple_Pack(1, op);
+            }
+            if (args == NULL) {
+                return -1;
+            }
+            pycomplex = PyComplex_Type.tp_new(&PyComplex_Type, args, NULL);
+            Py_DECREF(args);
+            if (pycomplex == NULL) {
+                return -1;
+            }
+            oop = PyComplex_AsCComplex(pycomplex);
+            Py_DECREF(pycomplex);
+            if (error_converting(oop.real)) {
+                return -1;
+            }
```

Comment:
```suggestion
            PyObject *args;
            if (PyBytes_Check(op)) {
                /* The complex constructor expects unicode */
                PyObject *unicode = PyUnicode_FromEncodedObject(op, NULL, NULL);
                if (unicode == NULL) {
                    return -1;
                }
                args = PyTuple_Pack(1, unicode);
                Py_DECREF(unicode);
            }
            else {
                args = PyTuple_Pack(1, op);
            }
            if (args == NULL) {
                return -1;
            }
            PyObject *pycomplex = PyComplex_Type.tp_new(&PyComplex_Type, args, NULL);
            Py_DECREF(args);
            if (pycomplex == NULL) {
                return -1;
            }
            oop = PyComplex_AsCComplex(pycomplex);
            Py_DECREF(pycomplex);
            if (error_converting(oop.real)) {
                return -1;
            }
```

Reply:
@eric-wieser  This just move the declaration of `pycomplex`, right?

Reply:
Yes, and `unicode`. Github is doing a lousy job of showing that diff.

Reply:
I prefer the shorter lines

Reply:
My ordered preference is:
```C
PyObject *unicode = PyUnicode_FromEncodedObject(op, NULL, NULL);
```
then
```C
PyObject *unicode =
    PyUnicode_FromEncodedObject(op, NULL, NULL);
```
then
```C
PyObject *unicode;
unicode = PyUnicode_FromEncodedObject(op, NULL, NULL);
```

At any rate, the C99 style is the first one, and these lines are not that long anyway.

Reply:
Dunno, I don't care much either way, part of the reason I did this was indeed longer than 80 char lines, plus I guess right now I don't like your second option too much, I guess just not used to it.